modelmapper/stats.py

```diff
@@ -0,0 +1,474 @@
+import decimal
+import datetime
+from collections import Counter
+from typing import Any, NamedTuple
+
+from modelmapper.types import (
+    HasBoolean,
+    HasDateTime,
+    HasDecimal,
+    HasDollar,
+    HasInt,
+    HasNull,
+    HasPercent,
+    HasString,
+)
+
+
+class FieldStats(NamedTuple):
+    counter: Any
+    max_int: 'FieldStats' = 0
+    max_pre_decimal: 'FieldStats' = 0
+    max_decimal_scale: 'FieldStats' = 0
+    max_string_len: 'FieldStats' = 0
+    datetime_formats: 'FieldStats' = None
+    len: 'FieldStats' = 0
+
+
+class UserInferenceRequired(ValueError):
+    """Thrown when we need the prompt the user for more information"""
+    def __init__(self, value_type, message):
+        super().__init__(message)
+        self.value_type = value_type
+
+
+class NoMatchFound(ValueError):
+    """Thrown when we can't match a datatype"""
+    pass
+
+
+class InconsistentData(ValueError):
+    """Thrown when we detect inconsistent data within a field"""
+    pass
+
+
+class TypeMatcher(object):
```

Comment:
Python 3 doesn't need you to subclass object.

Reply:
👍